### PR TITLE
ath79: add support for GL.iNet GL-X1200

### PR DESCRIPTION
**Pull request by dengxinfa**
This patch adds supports for GL-X1200.

Specification:
	- SOC: QCA9563 (775MHz)
	- Flash: 16 MiB
	- RAM: 128 MiB DDR2
	- Ethernet: 4x 1Gbps LAN + 1x 1Gbps WAN
	- Wireless: QCA9563(2.4GHz) and QCA9886(5GHz)
	- SIM: 2x SIM card slots
	- MicroSD: 1x microSD slot
	- Antenna: 2x external 5dBi antennas
	- USB: 1x USB 2.0 port
	- Button: 1x reset button
	- LED: 16x LEDs (3x GPIO controllable)
	- UART: 1x UART on PCB (JP1: 3.3V, RX, TX, GND)
	- OEM U-Boot supplies HTTP/GUI access

Implementation Notes
====================

Both the NOR and NAND variants boot off a NOR-based kernel,
consistent with the OEM's firmware.

The mode LEDs are
    * Boot, Running   system
    * Failsafe        2G
    * Upgrade         5G

Installation
============

Using sysupgrade
----------------

sysupgrade may be used to install a NAND image on a device running
a NAND image or a NOR image on a device running a NOR image. It is
recommended to *not* preserve config when upgrading from OEM firmware
or previous versions of OpenWrt. No supported sysupgrade path should
require "force". Transitioning from NOR to NAND can be accomplished

Using U-Boot
------------

The OEM U-Boot can be put into a graphical, firmware-upload mode by
holding down the button on the side of the router while applying power
and for a bit more than five seconds following with the current OEM
U-Boot. The power LED will come on, then the 5G LED will flash five
times, about once a second.  When the 5G LED stops flashing and the
2G LED lights solid, the router's U-Boot will provide an upload page
at http://192.168.1.1/ Either a browser may be used to upload an image,
or a utility such as curl may be used:

curl -X POST -F gl_firmware=\@*-nand-squashfs-factory.img \
         http://192.168.1.1/index.html
or
    curl -X POST -F gl_firmware=\@*-nor-squashfs-sysupgrade.bin \
         http://192.168.1.1/index.html

Note that NOR vs. NAND is based on the file name extension.

Signed-off-by: Xinfa Deng <xinfa.deng@gl-inet.com>

### CODE REVIEW

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 1]
```diff
@@ -25,6 +25,17 @@ case "$FIRMWARE" in
 		;;
 	esac
 	;;
+"ath10k/pre-cal-pci-0000:00:00.0.bin")
+	case $board in
+	glinet,gl-x1200-nor|\
+	glinet,gl-x1200-nor-nand)
+		caldata_extract "art" 0x5000 0x2f20
```
Comment (hauke) [inline target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:32]: Please define the position of the cal data in device tree. These commits are doing it for a different device: d4b3b23942336e22ef36ccbcf584efa6a55cc408 and 2d3321619b2b0a1eaabec3f47353e4e0aee3dc31.

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata [hunk 2]
```diff
@@ -25,6 +25,17 @@ case "$FIRMWARE" in
...
+		ln -sf /lib/firmware/ath10k/pre-cal-pci-0000\:00\:00.0.bin \
+			/lib/firmware/ath10k/QCA9888/hw2.0/board.bin
+		rm /lib/firmware/ath10k/QCA9888/hw2.0/board-2.bin
```
Comment (hauke) [inline target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata:35]: Why is this coping and removing needed?

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 1 above)
Comment (dengxinfa) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: hi hauke, I have modified according to your suggestion, thank you for your help

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 2 above)
Comment (dengxinfa) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: I have tested that ath10k cannot correctly identify the mac address in caldata <0x5006 0x6> without removing board-2.bin.
So I refer to comfast,cf-e313ac device, and remove board-2.bin.

**Comment by hauke** [comment]
Are the mac addresses correctly set on the interfaces (LAN and wifi) with the current changes in this pull request?

File: target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata (hunk 2 above)
Comment (hauke) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Thanks for the explanation. Is the mac address correctly set when you define the reference to the calibration data and mac address in the device tree like you did it now?

Comment (dengxinfa) [reply target/linux/ath79/nand/base-files/etc/hotplug.d/firmware/11-ath10k-caldata]: Yes, I have tested it and now the mac address is correct.

**Comment by dengxinfa** [comment]
> Are the mac addresses correctly set on the interfaces (LAN and wifi) with the current changes in this pull request?
Yes, all interfaces (lan and wifi) have got the mac address correctly.

**Comment by hauke** [comment]
Thank you for confirming that the MAC addresses are correct

**Comment by jkldgoefgkljefogeg** [comment]
How come there is still no build for GL-X1200?

https://downloads.openwrt.org/releases/23.05.2/targets/ath79/generic/

**Comment by hecatae** [comment]
> How come there is still no build for GL-X1200?
> 
> https://downloads.openwrt.org/releases/23.05.2/targets/ath79/generic/

https://downloads.openwrt.org/snapshots/targets/ath79/nand/ because it was added to Openwrt Master and is in Snapshots.

**Comment by jkldgoefgkljefogeg** [comment]
> > How come there is still no build for GL-X1200?
> > https://downloads.openwrt.org/releases/23.05.2/targets/ath79/generic/
> 
> https://downloads.openwrt.org/snapshots/targets/ath79/nand/ because it was added to Openwrt Master and is in Snapshots.

turned up there is a stable release for X1200, it's under https://downloads.openwrt.org/releases/23.05.2/targets/ath79/nand/ because X1200 has NOR + NAND?

**Pull request merged**